tests/c/outlinebase_bug.c

```diff
@@ -0,0 +1,79 @@
+// Run-time:
+//   env-var: YKD_PRINT_IR=jit-pre-opt
+//   env-var: YKD_SERIALISE_COMPILATION=1
+//   env-var: YKD_PRINT_JITSTATE=1
+//   stderr:
+//     ...
+//     --- Begin jit-pre-opt ---
+//     ...
+//     {{1}} = call i32 (ptr, ptr, ...) @fprintf(...
+//     ...
+//     {{2}} = call i32 (ptr, ptr, ...) @fprintf(...
+//     ...
+//     {{3}} = call i32 (ptr, ptr, ...) @fprintf(...
+//     ...
+//     {{4}} = call i32 (ptr, ptr, ...) @fprintf(...
+//     ...
+//     --- End jit-pre-opt ---
+//     ...
+//   stdout:
+//     exit
+
+// Test that outlining works as expected. Each trace in this test does two
+// rounds of the main interpreter loop, resulting in four `fprintf`s being
+// copied into the JIT module. However, if we forgot to reset the `OutlineBase`
+// in `jitmodbuilder.cc` after resetting `Outlining` to `false`, we will
+// continue outlining after seeing the first `yk_mt_control_point`, and miss
+// the second round of the loop.
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <yk.h>
+#include <yk_testing.h>
+
+int foo(int argc) {
+  int sum = 0;
+  // the loop ensures this function is outlined.
+  while(argc > 0) {
+    argc--;
+    sum = sum + argc;
+  }
+  return sum;
+}
+
+int bar(int argc) {
+  return foo(argc);
+}
+
+int main(int argc, char **argv) {
+  YkMT *mt = yk_mt_new(NULL);
+  yk_mt_hot_threshold_set(mt, 0);
+  YkLocation loc = yk_location_new();
+
+  int res = 9998;
+  int i = 6;
+  NOOPT_VAL(loc);
+  NOOPT_VAL(res);
+  NOOPT_VAL(i);
+  while (i > 0) {
+    int r = bar(argc + 2);
+    YkLocation *tmp;
+    if (i %2 == 0) {
+      tmp = &loc;
+    } else {
+      tmp = NULL;
+    }
+    yk_mt_control_point(mt, tmp);
+    fprintf(stderr, "r = %i\n", r);
```

Comment:
I don't think these two `fprintf`s are relevant to the test in the sense that their output isn't checked (let alone the inconsistent spaces either side of `=` :p)? Or, if they are, are they just `NOOPT_VAL` in disguise?

Reply:
It's just something I can easily check if it appears in the JIT IR. I agree the content of the print doesn't matter too much.

Reply:
I suggest we kill them -- it's just extra cruft from a reader's perspective.

Reply:
No, I need them to check that the JIT is doing the right thing. The test checks that there are four `fprintf`s in the trace. If there are only 2 then we've encountered the bug. But I can simplify them by passing the empty string or a single character as the print content is irrelevant.

Reply:
> But I can simplify them by passing the empty string or a single character as the print content is irrelevant.

Works for me!

Reply:
Try [136382e](https://github.com/ykjit/yk/pull/933/commits/136382e8ea14abd106a69545927b81dcbe975b40)